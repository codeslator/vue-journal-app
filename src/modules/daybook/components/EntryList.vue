<template>
  <div class="container-fluid entry-list">
    <div class="row">
      <div class="col">
        <input
          type="text"
          class="form-control"
          placeholder="Search entry"
          v-model="term"
        />
      </div>
    </div>
    <div class="row mt-2">
      <div class="col-12">
        <button
          class="btn btn-secondary btn-full"
          @click="$router.push({ name: 'entry', params: { id: 'new' } })"
        >
          <i class="fa fa-plus-circle"></i>
          New Entry
        </button>
      </div>
    </div>
    <div class="row mt-2">
      <div class="col-12">
        <div class="entry-list-scrollable">
          <entry-item
            v-for="entry in filteredEntries"
            :key="entry.id"
            :entry="entry"
          />
        </div>
      </div>
    </div>
  </div>
</template>

<script>
import { defineAsyncComponent } from "vue";
import { mapGetters } from "vuex";

export default {
  components: {
    EntryItem: defineAsyncComponent(() =>
      import(
        /* webpackChunkName: "DaybookEntryItem" */ "@/modules/daybook/components/EntryItem"
      )
    ),
  },
  data() {
    return {
      term: "",
    };
  },
  computed: {
    ...mapGetters("journal", {
      entries: "getEntriesByTerm",
    }),
    filteredEntries() {
      return this.entries(this.term);
    },
  },
};
</script>

<style lang="scss" scoped>
.entry-list {
  border-right: 1px dashed #34495e;
  height: 100%;
}

.entry-list-scrollable {
  overflow: scroll;
  height: calc(100vh - 115px);
  width: 100%;
  overflow-x: hidden;
}

.btn-full {
  width: 100%;
}
</style>