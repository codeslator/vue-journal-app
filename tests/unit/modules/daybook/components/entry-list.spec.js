import { shallowMount } from '@vue/test-utils';
import { createStore } from 'vuex';
import { daybookState } from '../../../mocks/daybook-state';
import journalStore from '@/modules/daybook/store/';
import EntryList from '@/modules/daybook/components/EntryList';

const createVuexStore = (initialState) => createStore({
  modules: {
    journal: {
      ...journalStore,
      state: { ...initialState }
    },
  }
});

describe('Entry List Component', () => {
  // const journalModuleMock = {
  //   namespaced: true,
  //   getters: {
  //     getEntriesByTerm,
  //   },
  //   state: () => ({
  //     isLoading: false,
  //     entries: daybookState.entries
  //   })
  // };

  // const store = createStore({
  //   modules: {
  //     journal: {
  //       ...journalModuleMock,
  //     }
  //   }
  // });

  const store = createVuexStore(daybookState);
  let wrapper;

  const mockRouter = {
    push: jest.fn(),
  };


  beforeEach(() => {
    jest.clearAllMocks();
    wrapper = shallowMount(EntryList, {
      global: {
        mocks: {
          $router: mockRouter,
        },
        plugins: [store]
      }
    });
  })

  test('Should match with last snapshot', () => {
    expect(wrapper.html()).toMatchSnapshot();
  })

  test('Should call getEntriesByTerm() without term and show 2 entries', () => {
    expect(wrapper.findAll('entry-item-stub').length).toBe(2);
  });

  test('Should to call getEntriesByTerm() and filter the entries', async () => {
    const input = wrapper.find('input');
    await input.setValue('mundo');
    expect(wrapper.findAll('entry-item-stub').length).toBe(1);
  });

  test('Button should to redirect to new', async () => {
    const button = wrapper.find('button');
    await button.trigger('click');
    expect(mockRouter.push).toHaveBeenCalledWith({ name: 'entry', params: { id: 'new' } });
  })
})