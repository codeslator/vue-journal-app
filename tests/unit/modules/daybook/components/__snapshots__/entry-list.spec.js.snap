// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Entry List Component Should match with last snapshot 1`] = `
<div class="container-fluid entry-list">
  <div class="row">
    <div class="col"><input type="text" class="form-control" placeholder="Search entry"></div>
  </div>
  <div class="row mt-2">
    <div class="col-12"><button class="btn btn-secondary btn-full"><i class="fa fa-plus-circle"></i> New Entry </button></div>
  </div>
  <div class="row mt-2">
    <div class="col-12">
      <div class="entry-list-scrollable">
        <entry-item-stub entry="[object Object]"></entry-item-stub>
        <entry-item-stub entry="[object Object]"></entry-item-stub>
      </div>
    </div>
  </div>
</div>
`;
